Feature/api controllers base

AlexBud369 opened the pull request:
## Changes

### Project Dependencies
- Added project reference to Infrastructure.csproj in Application.csproj to enable service dependencies
- Included Microsoft.AspNetCore.Http.Abstractions, Microsoft.IdentityModel.Tokens, and System.IdentityModel.Tokens.Jwt in Application.csproj for HTTP abstractions and JWT support
- Added Microsoft.IdentityModel.Tokens and System.IdentityModel.Tokens.Jwt to Infrastructure.csproj for JWT-related functionality in services
- Removed duplicate System.IdentityModel.Tokens.Jwt reference in API.csproj to clean up configuration

### Controllers

#### BooksController
- Added endpoints:
  - `GET /api/books` - List books
  - `GET /api/books/{id}` - Get book by ID
  - `POST /api/books` - Create book
  - `PUT /api/books/{id}` - Update book
  - `DELETE /api/books/{id}` - Delete book
  - `POST /api/books/{id}/cover` - Upload book cover
- Integrated MediatR, FluentValidation and Guard for GUID validation

#### FileController
- Implemented:
  - `DELETE /api/file/{fileName}` - File deletion with book access validation
  - `GET /api/file/{fileName}/signed-url` - Signed URL generation
- Uses IFileStorageService, ISignedUrlService and IBookAccessService

#### TagsController
- Added endpoints:
  - `GET /api/tags` - List tags
  - `GET /api/tags/{id}` - Get tag by ID
  - `POST /api/tags` - Create tag
  - `DELETE /api/tags/{id}` - Delete tag
- Integrated MediatR and FluentValidation

#### UsersController
- Implemented endpoints:
  - `GET /api/users` - List users
  - `GET /api/users/{id}` - Get user by ID
  - `GET /api/users/me` - Get current user
  - `PUT /api/users/{id}` - Update user
  - `DELETE /api/users/{id}` - Delete user
  - `POST /api/users/{id}/block` - Block user
  - `POST /api/users/{id}/role` - Change user role

#### AuthController
- Added endpoints for:
  - Registration
  - Login
  - Token refresh
  - Logout
- Validation via RegisterDtoValidator, LoginDtoValidator and RefreshTokenDtoValidator

### Services

#### FileStorageService
- Added IFileStorageService implementation for S3 file uploads and deletions

#### SignedUrlService
- Implemented ISignedUrlService for generating signed URLs for S3 file access

#### UserService
- Added to handle user updates and role retrieval
- Updated IUserService to include GetUserRolesAsync

#### AuthService
- Enhanced with:
  - Refresh token validation
  - Revocation logic via RevokeRefreshTokenAsync
  - Generation logic via RefreshTokenAsync

### Validators
- Added validators for:
  - DTOs: BookCreateDto, BookUpdateDto, BookFilterDto, UserUpdateDto, ChangeRoleRequestDto, RegisterDto, LoginDto, RefreshTokenDto
  - Commands/Queries: CreateBookCommand, UpdateBookCommand, DeleteTagCommand, ChangeRoleCommand, etc.
- Updated validators to use:
  - Flat localization keys
  - Domain constants (MinPageNumber, DefaultPageSize, SearchQueryMaxLength)
- Removed obsolete:
  - LoginCommandValidator
  - RegisterCommandValidator

### Command and Query Handlers
- Improved BookQueryService with:
  - Enum.Parse for sorting
  - Localization fixes
- Reworked TagService.AddTagsToBookAsync to:
  - Clear existing tags
  - Add new ones
  - Added GetTagNamesByIdsAsync
- Updated handlers:
  - CreateBookCommandHandler
  - UpdateBookCommandHandler
  - DeleteBookCommandHandler
  - For better tag handling and consistency
- Simplified:
  - RefreshTokenCommandHandler
  - LogoutCommandHandler
  - By moving logic to AuthService
- Fixed LoginCommandHandler to:
  - Use Guid for userId
  - Correct password validation logic

### Localization and Constants
- Updated SharedResources.en.resx
- Added SharedResources.ru.resx with Russian translations
- Added localization keys:
  - FailedToUpdateUser
  - InvalidRole
  - And others
- Introduced DomainConstants:
  - MinPageNumber
  - DefaultPageSize
  - SearchQueryMaxLength
  - For validator consistency

AlexBud369 committed:
refactore: remove useless files

AlexBud369 committed:
feat: add localization resources for authentication and signed URL generation

AlexBud369 committed:
feat: add AuthController for authentication endpoints
- Implemented AuthController with endpoints for registration, login, refresh-token, and logout

AlexBud369 committed:
fix: Update LogoutCommandHandler to return Unit

AlexBud369 committed:
feat: add validators for authentication DTOs
- Implemented RegisterDtoValidator, LoginDtoValidator, and RefreshTokenDtoValidator using FluentValidation

AlexBud369 committed:
feat: Add ISignedUrlService and SignedUrlService for S3 signed URLs

AlexBud369 committed:
fix: update LoginCommandHandler for correct type and validation
- Changed GenerateRefreshTokenAsync to use Guid instead of string for userId to match IAuthService signature
- Fixed Guard.AgainstFalse logic to correctly validate password, removing unnecessary negation

AlexBud369 committed:
feat: add RefreshTokenCommand and RefreshTokenCommandHandler
- Implemented RefreshTokenCommand and its handler to refresh JWT tokens using refresh tokens
- The handler validates the token, revokes the old token, generates new tokens via IAuthService, and updates the database using AppDbContext

AlexBud369 committed:
feat: simplify RefreshTokenCommandHandler and LogoutCommandHandler
- Moved refresh token and revoke token logic to AuthService.RefreshTokenAsync and AuthService.RevokeRefreshTokenAsync
- Updated RefreshTokenCommandHandler to call AuthService.RefreshTokenAsync with minimal validation
- Updated LogoutCommandHandler to call AuthService.RevokeRefreshTokenAsync with minimal validation
- Added LogoutCommand to define request structure.

AlexBud369 committed:
feat: enhance AuthService with token refresh and revocation logic
- Implemented RefreshTokenAsync to handle refresh token validation, revocation, and generation of new tokens
- Implemented RevokeRefreshTokenAsync to handle token revocation
- Fixed GenerateRefreshTokenAsync to use Guard.AgainstEmptyGuid for correct Guid validation
- Updated IAuthService to include RefreshTokenAsync and RevokeRefreshTokenAsync methods.

AlexBud369 committed:
fix: update SharedResources and localization files
- Removed duplicate keys and added FailedToUpdateUser key

AlexBud369 committed:
feat: add UpdateUserCommand, UpdateUserCommandHandler, UserUpdateDto and AutoMapper configuration
- Added UpdateUserCommand and UpdateUserCommandHandler for updating user details. Introduced UserUpdateDto for user update data transfer
- Configured AutoMapper mapping for UserUpdateDto to ApplicationUser, ignoring Id, IsBlocked, Books, and RefreshTokens, and setting default empty strings for Username and Email

AlexBud369 committed:
chore: remove LoginCommandValidator and RegisterCommandValidator
- Deleted LoginCommandValidator and RegisterCommandValidator as they are no longer needed, likely due to validation being handled in DTOs or services.

AlexBud369 committed:
feat: add validators for Book DTOs
- Added BookCreateDtoValidator, BookFilterDtoValidator, and BookUpdateDtoValidator to validate BookCreateDto, BookFilterDto, and BookUpdateDto respectively

AlexBud369 committed:
fix: update books command and query validators
- Updated CreateBookCommandValidator to validate CreateBookCommand instead of BookCreateDto. Fixed GetAllBooksQueryValidator to use flat localization keys  and DomainConstants.Book.SearchQueryMaxLength
- Updated UpdateBookCommandValidator to validate UpdateBookCommand instead of BookUpdateDto, added URL validation for CoverImageUrl, and fixed localization keys to flat structure
- Updated UploadBookCoverCommandValidator to rename BookId to Id, fix localization keys to flat structure, and add System.IO import

AlexBud369 committed:
fix: update tag validators to use flat localization keys
- Updated GetTagByIdQueryValidator to use TagIdRequired localization key
- Updated GetAllTagsQueryValidator to use flat localization keys and DomainConstants.Tag.MinPageNumber for PageNumber validation
- Updated CreateTagCommandValidator to use flat localization keys

AlexBud369 committed:
fix: update user validators
- Updated BlockUserCommandValidator, DeleteUserCommandValidator, GetAllUsersQueryValidator, and GetUserByIdQueryValidator to use flat localization keys
- Updated ChangeRoleCommandValidator to use flat localization keys and validate Role using Enum.TryParse with DomainConstants.User.AllowedRoles

AlexBud369 committed:
feat: add ChangeRoleRequestValidator
- Added ChangeRoleRequestValidator to validate ChangeRoleRequestDto with similar Role validation

AlexBud369 committed:
fix: update books command handlers
- Updated CreateBookCommandHandler to use GetTagNamesByIdsAsync for tag conversion before adding to book
- Fixed DeleteBookCommandHandler to return Task<Unit> for consistency with MediatR
- Updated UpdateBookCommandHandler to clear existing BookTag entries, handle null Tags, and rename _bookService to _bookAccessService for clarity

AlexBud369 committed:
feat: add DeleteTagCommand and DeleteTagCommandHandler
- Added DeleteTagCommand to define tag deletion request by ID
- Added DeleteTagCommandHandler to validate tag existence, ensure no associated books, and remove the tag from the database

AlexBud369 committed:
feat: add UserService and update IUserService
- Added UserService to handle user updates (username, email, password) with access validation and role retrieval. Updated IUserService to include GetUserRolesAsync method

AlexBud369 committed:
fix: enhance BookQueryService and TagService
- Improved BookQueryService by adding Enum.Parse for sortBy in ApplySorting with fallback to CreatedAt, and fixed localization to use _localizer
- Reworked TagService.AddTagsToBookAsync to clear existing BookTag entries, get or create tags, and add new BookTag entries
- Added TagService.GetTagNamesByIdsAsync to retrieve tag names by ID.

AlexBud369 committed:
feat: update localization resources and domain constants
- Updated SharedResources.en.resx to support {0} parameter in InvalidRole for allowed roles list. Added SharedResources.ru.resx with Russian translations for all keys
- Updated SharedResources.cs to reflect InvalidRole change. Added MinPageNumber to DomainConstants.Book and DefaultPageSize to DomainConstants.User for consistency with validators

AlexBud369 committed:
feat: add ChangeRoleRequestDto
- Added ChangeRoleRequestDto in Application.DTOs.User to support role change requests in UsersController for POST /api/users/{id}/role endpoint

AlexBud369 committed:
feat: add Application.Interfaces namespace to FileStorageService

AlexBud369 committed:
feat: add BooksController
- Added BooksController in API.Controllers to handle book-related operations
- Implemented endpoints for listing books (GET /api/books), retrieving a book by ID (GET /api/books/{id}), creating a book (POST /api/books), updating a book (PUT /api/books/{id}), deleting a book (DELETE /api/books/{id}), and uploading a book cover (POST /api/books/{id}/cover)
- Integrated MediatR for command/query handling, FluentValidation for input validation, and Guard for GUID validation

AlexBud369 committed:
feat: add FileController
- Added FileController in API.Controllers to handle file operations
- Implemented endpoints for deleting a file (DELETE /api/file/{fileName}) with book access validation and generating a signed URL (GET /api/file/{fileName}/signed-url)
- Integrated IFileStorageService, ISignedUrlService, IBookAccessService, UserManager, and AppDbContext with Guard for input validation

AlexBud369 committed:
feat: add TagsController
- Added TagsController in API.Controllers to handle tag-related operations
- Implemented endpoints for listing tags (GET /api/tags), retrieving a tag by ID (GET /api/tags/{id}), creating a tag (POST /api/tags), and deleting a tag (DELETE /api/tags/{id})
- Integrated MediatR for command/query handling, FluentValidation for input validation, and Guard for GUID validation

AlexBud369 committed:
feat: add UsersController
- Added UsersController in API.Controllers to handle user-related operations
- Implemented endpoints for listing users (GET /api/users), retrieving a user by ID (GET /api/users/{id}), getting the current user (GET /api/users/me), updating a user (PUT /api/users/{id}), deleting a user (DELETE /api/users/{id}), blocking a user (POST /api/users/{id}/block), and changing a user role (POST /api/users/{id}/role)
- Integrated MediatR for command/query handling, FluentValidation for input validation, and Guard for GUID validation

AlexBud369 committed:
feat: update project dependencies for Application, Infrastructure, and API
- Added project reference to Infrastructure.csproj in Application.csproj to enable service dependencies
- Added Microsoft.AspNetCore.Http.Abstractions, Microsoft.IdentityModel.Tokens, and System.IdentityModel.Tokens.Jwt packages to Application.csproj for HTTP abstractions and JWT support
- Added Microsoft.IdentityModel.Tokens and System.IdentityModel.Tokens.Jwt packages to Infrastructure.csproj for JWT-related functionality in services
- Removed duplicate System.IdentityModel.Tokens.Jwt package reference in API.csproj to clean up project configuration

AlexBud369 committed:
fix: update pagination constants in GetAllTagsQuery and GetAllUsersQuery

AlexBud369 committed:
fix: correct IBookQueryService and ITagQueryService namespace location

AlexBud369 committed:
feat: add validators for DeleteTagCommand, UpdateUserCommand, and UserUpdateDto

AlexBud369 committed:
feat: create Persistence project to resolve cyclic dependency
- Added new Persistence class library project to house AppDbContext and data access components
- Configured Persistence.csproj with .NET 9.0 and Entity Framework Core 9.0.0 packages
- Added project references to Persistence in API.csproj, Application.csproj, and Infrastructure.csproj

AlexBud369 committed:
refactor: move AppDbContext from Infrastructure to Persistence
- Relocated AppDbContext.cs from Infrastructure/Data to Persistence.
- Updated namespace in AppDbContext.cs from Infrastructure.Data to Persistence.
- Removed Infrastructure/Data folder as data access is now handled by Persistence project.

AlexBud369 committed:
refactor: update FileController and UsersController
- Removed Guard.Initialize(_localizer) from DeleteFile method in FileController.cs, as initialization is handled in the constructor
- Added private readonly IValidator<UpdateUserCommand> _updateUserCommandValidator field to UsersController.cs
- Updated UsersController constructor to include IValidator<UpdateUserCommand> and integrated validator in UpdateUser method

AlexBud369 committed:
refactor: replace Infrastructure.Data with Persistence in services and handlers

AlexBud369 committed:
feat: add Persistence project references to API, Infrastructure, and Application
- Added MediatR (11.1.0), MediatR.Extensions.Microsoft.DependencyInjection (11.1.0), and Microsoft.AspNetCore.Authentication.JwtBearer (9.0.0) to API.csproj
- Added project reference to Persistence in API.csproj
- Added project reference to Persistence in Infrastructure.csproj, retaining EF Core and Identity packages as changes predate Persistence creation
- Replaced Infrastructure project reference with Persistence in Application.csproj

AlexBud369 committed:
feat: update Program.cs with full service configuration and Persistence namespace
- Updated Program.cs to include MediatR, AutoMapper, Localization, AWS S3, services, and validators
- Replaced Infrastructure.Data with Persistence namespace for AppDbContext
- Added JWT authentication, authorization, and controller configuration

AlexBud369 committed:
refactor: remove Application.Common.Resources from ErrorHandlingMiddleware

AlexBud369 committed:
feat: add pagination constants and UserRole enum to DomainConstants

AlexBud369 committed:
feat: update command classes to use IRequest<Unit> for MediatR

AlexBud369 committed:
feat: add initial database migration for BookSharingPlatform
- Created InitialCreate migration for AspNetUsers, AspNetRoles, Books, Tags, BookTag, RefreshTokens
- Added AppDbContext configuration for entities
- Generated model snapshot for EF Core 9.0.0

AlexBud369 committed:
chore: update csproj files with new packages and version updates

AlexBud369 committed:
feat: add AppDbContextFactory for EF Core migrations
- Implemented IDesignTimeDbContextFactory<AppDbContext> to enable migrations
- Configured factory to load appsettings.json from API project

AlexBud369 committed:
fix: address nullability warnings and improve localization
- Fixed CS8602 in TagService.cs by adding null check in ValidateInputs
- Fixed CS8602/CS8604 in FileController.cs by replacing Guard.AgainstNull with explicit null check
- Improved Guard.cs to safely handle null localizer with fallback messages
- Added FailedToUpdateUser resource to SharedResources.cs
- Removed unused using from UserService.cs

AlexBud369 committed:
fix: correct database connection string

AlexBud369 committed:
fix: Improve localization and configuration handling
- Updated Guard.cs to safely handle null localizer with fallback messages
- Added FailedToUpdateUser resource to SharedResources.cs
- Added Jwt:Key validation in Program.cs
- Removed unused using Application.Common.Resources from FileStorageService.cs

budnikme commented on src/API/Controllers/AuthController.cs:
```diff
@@ -0,0 +1,111 @@
+            return BadRequest(validationResult.Errors);
+        }
+
+        var command = new RegisterCommand
```
Please store command/query and handler in the same class
```
public static class RefreshToken
{
    public class Command : IRequest<AuthResponseDto>
    {

    }

    public class Handler : IRequestHandler<Command, AuthResponseDto>
    {
        public Task Handle()
        {
        }
    }
}
```

budnikme commented on src/API/Controllers/FileController.cs:
```diff
@@ -0,0 +1,81 @@
+        Guard.AgainstEmptyString(fileName, nameof(fileName), _localizer.GetString(SharedResources.FileNameRequired));
+        Guard.AgainstEmptyGuid(bookId, nameof(bookId), _localizer.GetString(SharedResources.BookIdRequired));
+
+        var user = await _userManager.GetUserAsync(User);
```
Please avoid implementing any business logic inside controllers

AlexBud369 committed:
refactor: consolidate Login command and handler into single file
- Combined LoginCommand and LoginCommandHandler into Login.cs
- Removed redundant validation (Email, Password) from handler, relying on LoginDtoValidator
- Split Handle method into sub-methods for better readability
- Replaced magic string 'Role' with DomainConstants.Mapper.RoleKey

AlexBud369 committed:
refactor: Consolidate Register command and handler into single file
- Combined RegisterCommand and RegisterCommandHandler into Register.cs as per reviewer feedback
- Removed redundant validation (Email, Password, UserName) from handler, relying on RegisterDtoValidator
- Split Handle method into sub-methods for better readability
- Used UserRole.User instead of UserRole.User.ToString() in CreateUserAsync for type safety

AlexBud369 committed:
refactor: consolidate RenewToken and Logout commands and handlers
- Combined RefreshTokenCommand and RefreshTokenCommandHandler into RenewToken.cs, renamed to avoid naming conflict with Domain.Entities.RefreshToken
- Combined LogoutCommand and LogoutCommandHandler into Logout.cs as per reviewer feedback
- Added Guard.AgainstNull for RefreshTokenAsync result in RenewToken.cs for robustness
- Kept existing validation in Logout.cs as it complements LogoutCommandValidator

AlexBud369 committed:
refactor: update AuthController and validators for consistency
- Updated AuthController.cs to reference new command classes (Login, Register, RenewToken, Logout)
- Replaced magic string 'sub' with DomainConstants.Jwt.ClaimSub in Logout
- Added localization to Unauthorized response in Logout
- Added IValidator<Logout.Command> to AuthController constructor
- Added PasswordMaxLength validation to LoginDtoValidator
- Added LogoutCommandValidator to validate Logout.Command

AlexBud369 committed:
feat: update BookMappingProfile and BookFilterDto
- Added Id and CreatedAt mappings in BookMappingProfile for BookCreateDto to Book
- Updated BookFilterDto to use DomainConstants.Book.DefaultPageNumber and DefaultPageSize

AlexBud369 committed:
feat: consolidate and simplify book command handlers
- Merged CreateBookCommand, UpdateBookCommand, DeleteBookCommand, and UploadBookCoverCommand with their handlers
- Simplified handler logic for book creation, update, deletion, and cover upload
- Ensured usage of DomainConstants for validation (e.g., Book.TitleMaxLength, BookCover.MaxFileSizeBytes)
- Integrated IBookAccessService and IImageService for access checks and image handling

AlexBud369 committed:
feat: consolidate and simplify book query handlers
- Merged GetAllBooksQuery and GetBookByIdQuery with their handlers
- Simplified query logic for book retrieval with filtering and pagination
- Ensured usage of DomainConstants.Book for pagination (DefaultPageNumber, DefaultPageSize, MaxPageSize)
- Integrated IBookAccessService for access checks in GetAllBooksQuery

AlexBud369 committed:
feat: add IJwtTokenService and JwtTokenService

AlexBud369 committed:
feat: update UsersController with GetCurrentUserId
- Added GetCurrentUserId method to UsersController to reduce code duplication
- Ensured compatibility with DomainConstants.User for pagination and SharedResources for error handling

AlexBud369 committed:
feat: refactor IAuthService and AuthService
- Updated IAuthService to use UserRole instead of string for CreateUserAsync
- Refactored AuthService to delegate JWT generation to IJwtTokenService
- Used DomainConstants.Jwt.JwtRefreshTokenLifetimeDays for refresh token expiry
- Used DomainConstants.Mapper.RoleKey for role mapping in AuthResponseDto
- Added validation with SharedResources.JwtTokenGenerationFailed and other error keys
- Simplified RefreshTokenAsync with helper methods for validation and token generation

AlexBud369 committed:
feat: enhance IBookAccessService and BookAccessService
- Added CanDeleteBookAsync and CanEditBookAsync to IBookAccessService
- Implemented CanDeleteBookAsync and CanEditBookAsync in BookAccessService with user and admin checks
- Added AppDbContext and UserManager dependencies for data access and role checks
- Used SharedResources for validation (BookNotFound, BookIdRequired, UserIdRequired, UnauthorizedAccess)
- Ensured ValidateBookAccessAsync returns Task for consistency

AlexBud369 committed:
feat: enhance IBookQueryService and BookQueryService
- Added GetBooksQueryAsync to IBookQueryService for filtered IQueryable<Book>
- Implemented GetBooksQueryAsync in BookQueryService using BookFilterDto
- Updated GetBooksAsync to use TotalItems in PagedResponseDto
- Used SharedResources for validation (FilterRequired, BookIdRequired, BookNotFound)
- Added necessary namespaces for BookSortBy and BookFilterDto

AlexBud369 committed:
feat: enhance IImageService and ImageService
- Added UploadImageAsync to IImageService for image uploads
- Implemented UploadImageAsync in ImageService with validation for size and extension
- Used DomainConstants.BookCover.MaxFileSizeBytes and AllowedImageExtensions
- Used SharedResources for validation (CoverImageRequired, CoverImageEmpty, CoverImageTooLarge, InvalidFileExtension, CoverImageUploadFailed)
- Simplified validation logic in UpdateBookCoverAsync

AlexBud369 committed:
feat: enhance ITagService and TagService
- Added UpdateTagsForBookAsync to ITagService for updating book tags
- Implemented UpdateTagsForBookAsync in TagService with tag clearing and addition
- Used SharedResources for validation (BookNotFound, BookIdRequired, TagsRequired, TagNameRequired)
- Optimized handling of empty tag lists in UpdateTagsForBookAsync

AlexBud369 committed:
feat(tags): combine CreateTag and DeleteTag with handlers
- Added Guard.AgainstMaxLength for TagName validation with DomainConstants.Tag.NameMaxLength
- Confirmed mapping with TagMappingProfile
- Ensured tag deletion checks for associated books

AlexBud369 committed:
feat: combine GetAllTags and GetTagById with handlers and PagedResponseDto
- Merged GetAllTagsQuery and GetTagByIdQuery with their handlers into single classes
- Updated ITagQueryService and TagQueryService to return PagedResponseDto<TagDto> for GetAllTags
- Used DomainConstants.Tag for DefaultPageNumber, DefaultPageSize, MinPageNumber, MaxPageSize
- Validated with SharedResources.InvalidPageNumber, InvalidPageSize, TagIdRequired, TagNotFound
- Removed separate query and handler files

AlexBud369 committed:
feat: rename validators for CreateTag and GetAllTags
- Renamed CreateTagCommandValidator and GetAllTagsQueryValidator to match new CreateTag and GetAllTags classes
- Maintained validation with DomainConstants.Tag.NameMaxLength, MinPageNumber, MaxPageSize
- Used SharedResources for TagNameRequired, TagNameTooLong, InvalidTagNameFormat, InvalidPageNumber, InvalidPageSize
- Ensured consistent validation logic

AlexBud369 committed:
feat: update TagsController to use DomainConstants
- Replaced magic numbers in GetTags with DomainConstants.Tag.DefaultPageNumber and DefaultPageSize
- Updated validator types to match new CreateTag, GetAllTags, GetTagById, and DeleteTag classes
- Ensured no business logic and maintained small method sizes

AlexBud369 committed:
feat: Update book validators to use Command/Query structure

AlexBud369 committed:
feat: Update user validators to use Command/Query structure
- Removed ChangeRoleRequestValidator

AlexBud369 committed:
feat: update user commands with Command/Handler structure
- Updated BlockUser to use BlockUserAsync
- Structured BlockUser, ChangeRole, DeleteUser, UpdateUser as static classes with Command and Handler
- Validated via respective validators
- Maintained IUserService integration

AlexBud369 committed:
feat: update user queries with Query/Handler structure
- Structured GetAllUsers and GetUserById as static classes with Query and Handler
- Returns PagedResponseDto<UserDto> for GetAllUsers and UserDto for GetUserById
- Validated via GetAllUsersValidator and GetUserByIdValidator
- Integrated with IUserQueryService"

AlexBud369 committed:
feat: enhance UserService and UserQueryService with new functionality
- Added BlockUserAsync, EnsureIsAdminAsync, GetUserByIdAsync, ChangeUserRoleAsync, DeleteUserAsync, MapToUserDtoAsync to UserService
- Refactored UpdateUserAsync to use dedicated methods (EnsureCanUpdateAsync, UpdateUsernameAsync, UpdateEmailAsync, UpdatePasswordAsync)
- Added RoleManager and IMapper dependencies to UserService
- Updated UserQueryService to return PagedResponseDto<UserDto> in GetUsersAsync with role mapping
- Added UserManager dependency to UserQueryService for role retrieval
- Made email and username searches case-insensitive in UserQueryService
- Updated IUserService and IUserQueryService interfaces to reflect new methods and PagedResponseDto
- Replaced BlockUserCommand with BlockUser.Command in BlockUserValidator

AlexBud369 committed:
feat:update UsersController for new command/query structure
- Updated validators to use UpdateUser, DeleteUser, BlockUser.Command, ChangeRole, GetAllUsers.Query, GetUserById.Query
- Removed ChangeRoleRequestDto validator and integrated role validation into ChangeRoleValidator
- Replaced hardcoded role 'Admin' with UserRole.Admin.ToString()
- Used DomainConstants.User for default pageNumber and pageSize
- Simplified ChangeRole method by removing separate ChangeRoleRequestDto validation
- Maintained ValidateAndCreateCommand for command validation

AlexBud369 committed:
feat: update BooksController and TagsController for new command/query structure
- Replaced User.FindFirst('sub') with DomainConstants.Jwt.ClaimSub in BooksController
- Updated role checks to use nameof(UserRole.Admin) in BooksController and TagsController
- Replaced hardcoded pagination values with DomainConstants.Tag.DefaultPageNumber and DefaultPageSize in TagsController
- Maintained validation and MediatR integration

AlexBud369 committed:
feat: update Program.cs for new command/query structure
- Updated validator registrations for UpdateUser, DeleteUser, BlockUser.Command, ChangeRole, GetAllUsers.Query, GetUserById.Query
- Updated book validator registrations for CreateBook.Command, UpdateBook.Command, DeleteBook.Command, GetAllBooks.Query, GetBookById.Query, UploadBookCover.Command
- Removed ChangeRoleRequestValidator registration
- Added DeleteBookCoverValidator registration
- Updated MediatR registration to use CreateBook assembly

AlexBud369 committed:
feat: refactor FileController to remove business logic
- Moved DeleteFile logic to DeleteBookCover.Command and Handler
- Added DeleteBookCoverValidator for FileName, BookId, UserId validation
- Replaced magic numbers with constants
- Removed IFileStorageService, IBookAccessService, UserManager, AppDbContext dependencies from FileController
- Used DomainConstants.Jwt.ClaimSub for userId retrieval

AlexBud369 committed:
feat: Add localization keys and constants for file operations

budnikme commented on src/Application/Services/AuthService.cs:
```diff
@@ -126,41 +80,66 @@ public async Task RevokeRefreshTokenAsync(string refreshToken, Guid userId, Canc
+
+        var refreshToken = await _context.RefreshTokens
+            .Include(rt => rt.User)
+            .FirstOrDefaultAsync(rt => rt.Token == token && !rt.IsRevoked && rt.ExpiresAt > DateTime.UtcNow, cancellationToken);
```
Please move that to the predicate expression https://medium.com/@omermemes83/understanding-expression-func-t-bool-with-and-without-ef-core-70dc218e7a59

budnikme commented on src/Application/Features/Books/Commands/CreateBook.cs:
```diff
@@ -0,0 +1,98 @@
+
+        private async Task AddTagsAsync(Book book, IEnumerable<Guid>? tagIds, CancellationToken cancellationToken)
+        {
+            if (tagIds != null && tagIds.Any()) {
```
Please invert this if expression
```
if (!tagIds?.Any()){
    return;
}
```

budnikme commented on src/Application/Services/UserQueryService.cs:
```diff
@@ -1,59 +1,82 @@
-        return _mapper.Map<IEnumerable<UserDto>>(users);
+        var userDtos = new List<UserDto>();
+        foreach (var user in users) {
+            var roles = await _userManager.GetRolesAsync(user);
```
Such await inside a for loop can cause performance issues in the future

AlexBud369 committed:
feat: add ExpressionExtensions for dynamic predicate combination
- Created ExpressionExtensions with And and Or methods to combine Expression<Func<T, bool>> predicates
- Supports dynamic query building for future extensibility, although not used in current implementation

AlexBud369 committed:
feat: add BookId and TagId constants to DomainConstants

AlexBud369 committed:
fix(services): optimize services by removing await in loops and improving code readability
- Replaced await in loop with single query in UserQueryService.GetUsersAsync using Join to fetch roles, resolving N+1 issue
- Replaced await in loop with AddRangeAsync in TagService.AddBookTagsAsync to add multiple BookTag entries in a single query (previously done)
- Used DomainConstants.Book.BookId and DomainConstants.Tag.TagId in TagService.AddBookTagsAsync and ClearExistingBookTagsAsync
- Used dynamic predicates with Expression<Func<RefreshToken, bool>> in AuthService.RevokeRefreshTokenAsync and ValidateRefreshTokenAsync
- Inverted if expression in AuthService.CreateUserAsync for better readability
- Applied if inversion in ImageService.DeleteBookCoverAsync
- Improved Guard checks readability and used constants in UserService
- Replaced 'Admin' string with UserRole.Admin.ToString() in BookAccessService

AlexBud369 committed:
fix: improve readability and validation in CreateBook, DeleteBook, and UploadBookCover
- Inverted if expressions in CreateBook.AddTagsAsync, DeleteBook.Handle for better readability
- Removed redundant Guard.AgainstNull checks in CreateBook
- Added tagNames.Any() validation in CreateBook.AddTagsAsync

budnikme commented on src/Application/Features/Books/Commands/DeleteBook.cs:
```diff
@@ -37,7 +37,10 @@ public async Task Handle(Command request, CancellationToken cancellationToken)
             ValidateRequest(request);
             var book = await GetBookAsync(request.Id, cancellationToken);
-            await ValidateAccessAsync(book, request.UserId, cancellationToken);
+            if (!await _bookAccessService.CanDeleteBookAsync(request.UserId, book.Id, cancellationToken)){
```
This will look much clearer
```
var canDelete = bookAccessService.CanDelete(...);
Guard.AgainstFalse(canDelete);
```
I suppose that your guard will throw an error only if false value is passed

budnikme commented on src/Application/Services/UserQueryService.cs:
```diff
@@ -63,12 +64,25 @@ public async Task<PagedResponseDto<UserDto>> GetUsersAsync(
-            var userDto = _mapper.Map<UserDto>(user, opts => opts.Items["Role"] = roles.FirstOrDefault() ?? UserRole.User.ToString());
-            userDtos.Add(userDto);
-        }
+        var userIds = users.Select(u => u.Id).ToList();
```
It is recommended to always use UserManager instead of own implementation. I understand that you have done that to avoid for loop and enumerating over multiple users, but is there even a need to get roles of multiple users at the same time?

AlexBud369 committed:
fix: remove role fetching from UserQueryService.GetUsersAsync
- Removed role fetching from GetUsersAsync to avoid N+1 and use UserManager as recommended
- Simplified mapping in GetUsersAsync

AlexBud369 committed:
fix: improve readability in DeleteBook
- Replaced inverted if with clear canDelete check and Guard.AgainstFalse for UnauthorizedAccess
